dandi/dandiapi.py

```diff
@@ -620,3 +598,45 @@ def delete_asset_bypath(self, dandiset_id, version_id, asset_path):
         if asset is None:
             raise RuntimeError(f"No asset found with path {asset_path!r}")
         self.delete_asset(dandiset_id, version_id, asset["asset_id"])
+
+
+def upload_part(storage_session, fp, lock, etagger, asset_path, part):
+    etag_part = etagger.get_part(part["part_number"])
+    if part["size"] != etag_part.size:
+        raise RuntimeError(
+            f"Server and client disagree on size of upload part"
+            f" {part['part_number']}; server says {part['size']},"
+            f" client says {etag_part.size}"
+        )
+    with lock:
+        fp.seek(etag_part.offset)
+        chunk = fp.read(part["size"])
+    if len(chunk) != part["size"]:
+        raise RuntimeError(f"End of file {fp.name} reached unexpectedly early")
```

Comment:
while at it (although just an older code moved):
```suggestion
    if len(chunk) != part["size"]:
        raise RuntimeError(f"End of file {fp.name} reached unexpectedly early: read {len(chunk)} from expected {part['size']}")
```

Reply:
Done.